src/Simulator.sol

```diff
@@ -125,6 +126,90 @@ contract Simulator {
         return _callOrchestrator(oc, isStateOverride, data);
     }
 
+    /// @dev Performs a call to Multicall3 with pre-calls followed by an orchestrator call.
+    /// Returns the gas used parsed from the last Result in the multicall3 response.
+    /// If parsing fails (gasUsed == 0), this function stores the orchestrator's error in memory
+    /// so the caller can bubble it up using bubbleUpMulticall3Error.
+    /// @param multicall3 The multicall3 contract address
+    /// @param preCalls Array of Call3 structs representing calls to execute before the orchestrator call
+    /// @param oc The orchestrator address
+    /// @param isStateOverride Whether to use state override mode for the orchestrator call
+    /// @param combinedGasOverride The combined gas override value
+    /// @param u The Intent struct to pass to the orchestrator
+    function _callMulticall3(
+        address multicall3,
+        IMulticall3.Call3[] memory preCalls,
+        address oc,
+        bool isStateOverride,
+        uint256 combinedGasOverride,
+        ICommon.Intent memory u
+    ) internal freeTempMemory returns (uint256 gasUsed, bytes memory lastReturnData) {
+        // Build the orchestrator call data
+        bytes memory orchestratorData = abi.encodeWithSignature(
+            "simulateExecute(bool,uint256,bytes)",
+            isStateOverride,
+            combinedGasOverride,
+            abi.encode(u)
+        );
+
+        // Construct the full Call3[] array: preCalls + orchestrator call
+        IMulticall3.Call3[] memory calls = new IMulticall3.Call3[](preCalls.length + 1);
+        for (uint256 i = 0; i < preCalls.length; i++) {
+            calls[i] = preCalls[i];
+        }
+        // Last call is to the orchestrator
+        calls[preCalls.length] =
+            IMulticall3.Call3({target: oc, allowFailure: true, callData: orchestratorData});
+
+        // Call multicall3.aggregate3
+        IMulticall3.Result[] memory results = IMulticall3(multicall3).aggregate3(calls);
+
+        // Get the last result (orchestrator call result)
+        // Check all preCalls for failures (all results except the last one, which is the orchestrator call)
+        if (results.length > 1) {
+            for (uint256 i = 0; i < results.length - 1; i++) {
+                // If any pre-call failed, we return gasUsed = 0 and the error data from that call
+                if (!results[i].success) {
+                    return (0, results[i].returnData);
+                }
+            }
+        }
+        IMulticall3.Result memory lastResult = results[results.length - 1];
+        lastReturnData = lastResult.returnData;
+
+        // Parse gasUsed from the orchestrator's return data
```

Comment:
this is only returning the `gasUsed` by the intent right? would be interested in the overal gasUsed of the whole multicall including the intent and precalls

Reply:
actually both. used by intent (so we can cap its execution) and by intent+precalls so we can price it